README.md
RIchCoin [XRI] Source Tree
================================
Auspicium Melioris Aevi

```
|_   _| |__   ___   / ___|___ (_)_ __     ___  / _|   __ _  | __ )  ___| |_| |_ ___ _ __   / \   __ _  ___
  | | | '_ \ / _ \ | |   / _ \| | '_ \   / _ \| |_   / _` | |  _ \ / _ \ __| __/ _ \ '__| / _ \ / _` |/ _ \
  | | | | | |  __/ | |__| (_) | | | | | | (_) |  _| | (_| | | |_) |  __/ |_| ||  __/ |   / ___ \ (_| |  __/
  |_| |_| |_|\___|  \____\___/|_|_| |_|  \___/|_|    \__,_| |____/ \___|\__|\__\___|_|  /_/   \_\__, |\___|
                                                                                                |___/    
                                                                                                
                                                                                 2018 RIchCoin/XRI
```

Specifications:
--------------

* Symbol: `XRI`
* PoW (proof of work)
* Algorithms: scrypt, x17, Lyra2rev2, myr-groestl, & blake2s
* Blocktime: 60 seconds
* RPC port: `20203`
* P2P port: `21203`

* Block Reward & Circulating Supply(CR) and Time Period (TP) in months after release:
  * Block 0 to 44,000 : 1440 coins (CR 63,360,000 TP 0-1)
  * 44,000 to 132,000 : 720 coins (CR 158,400,000 TP 1-4)
  * 132,000 to 440,000: 360 coins (CR 253,440,000 TP 4-10)
  * 440,000 to 968,000: 180 coins (CR 348,480,000 TP 10-22)
  * 968,000 to 1,760,000: 90 coins (CR 419,760,000 TP 22-40)
  * 1,760,000 to 2,816,000: 45 coins (CR 467,280,000 TP 40-64)
  * 2,816,000 to 4,224,000: 24 coins (CR 499,664,000 TP 64-96)

* After 8 years, Block Rewards will be dictated by network hashrates(HR) as such:
  * HR 16x: Block Reward 0.6x (CR 500,000,000+ TP 96-)
  * HR 32/: Block Reward 1.4x (CR 500,000,000+ TP 96-)

Total Supply
------------
After 8 years, by mid-2026, circulating supply will be 499,664,000 coins (approx. 500 million XRI), before automation of the block reward begins. Current maximum set is 2 billion, but this can be changed anytime.

Using RIchCoin on Windows (To be tested & added)
-------------
1. Download the pre-compiled software from the latest release in [this repository](https://github.com/RIchCoin-devs/RIchCoin-WindowsWallet/releases).
2. Install the software.
3. In Windows File Explorer, open c:\Users\XXX\AppData\Roaming\RIchCoin (be sure to change XXX to your windows user)
4. Right click and create a new file RIchCoin.txt
5. Edit the file to have contents above (see instructions above for options)
6. Save and close the file
7. Reame the file to RIchCoin.conf.
8. Start the RIchCoin-qt program.
9. Open up RIchCoin-qt console and run ```getinfo``` (or ```getmininginfo```) to verify settings.
10. Proceed below to make changes to your RIchCoin.conf file.
**Note:** You must re-start the wallet after making changes to RIchCoin.conf.

Compiling Linux Wallet on Ubuntu/Debian
----------------------

1. Install the dependencies.

```sudo apt-get update```

```sudo apt-get install libdb-dev libdb++-dev build-essential libtool autotools-dev automake pkg-config libssl1.0-dev libevent-dev bsdmainutils git libboost-all-dev libminiupnpc-dev libqt5gui5 libqt5core5a libqt5webkit5-dev libqt5dbus5 qttools5-dev qttools5-dev-tools libprotobuf-dev protobuf-compiler libqrencode-dev```

**Note**: If you are on debian, you will also need to `apt-get install libcanberra-gtk-module`.

2. Clone the git repository and compile the daemon and gui wallet:

```git clone https://github.com/HTY2003/RIchCoin && cd RIchCoin && ./autogen.sh && ./configure --with-incompatible-bdb && make```

if you are using source-build libdb4.8(++)-dev you may need to use

```./configure  CPPFLAGS="-I/usr/local/BerkeleyDB.4.8/include -O2" LDFLAGS="-L/usr/local/BerkeleyDB.4.8/lib"```

**Note**: If you get a "memory exhausted" error, make a swap space on your boot drive.

Compiling Mac Wallet on MacOS (To be tested):
------------
1. Ensure you do not have qt5 nor qt installed.

    `brew install qt@5.5 boost@1.57 automake miniupnpc openssl protobuf protobuf berkeley-db@4`   
    `brew link boost@1.57 --force`


2. Download the wallet source and build:

    `git clone https://github.com/HTY2003/RIchCoin`

    `cd RIchCoin`

    `./building/common.sh`

    `./building/build.sh`

    `./building/dist.sh`


Building the Mac installer (.dmg) file
-----
Run `make deploy`

If you are building the .dmg (by running 'mac deploy') you may need to run these commands if you get an error regarding mysql:

    brew install mysql

    cd /usr/local/qt5/5.4/clang_64/plugins/sqldrivers

    otool -L libqsqlmysql.dylib

Note: This may be pointing to an version of mysql that you do not have installed (like mysql55) - Alternatively, you may be able to remove the library from the sqldrivers folder.

    install_name_tool -change /opt/local/lib/mysql55/mysql/libmysqlclient.18.dylib /usr/local/Cellar/mysql/5.7.12/lib/libmysqlclient.20.dylib libqsqlmysql.dylib

Note: You may also run into issues when using `macdeployqtplus` to create the bundle, and the library will not bundle all of the boost dylibs. It's highly recommended to use the functions provided inside of [dylib-fixes.sh](/building/mac/dylib-fixes.sh)

Trying to build .dmg on 10.8? You will need to run this:

    export CFLAGS=-Qunused-arguments

    export CPPFLAGS=-Qunused-arguments

    sudo -E easy_install appscript
    
Accessing the wallet (Linux & Mac):
----
The gui wallet is in ./RIchCoin/src/qt and the daemon(background) wallet in ./RIchCoin/src directories.

**Note**: If you see something like 'Killed (program cc1plus)' run ```dmesg``` to see the error(s)/problems(s). This is most likely caused by running out of resources. You may need to add some RAM or add some swap space.

**Optional**:
If you want to copy the binaries for use by all users, run the following commands:

```sudo cp src/RIchCoind /usr/bin/```

```sudo cp src/qt/RIchCoin-qt /usr/bin/```

Creating a Configuration File
------------------------------

Type ```cd ~``` to get back to the home folder and type:

```RIchCoind.exe``` (for Windows users)
```./RIchCoind``` (for Mac and Linux users)

The output from this command will tell you that you need to make a Richcoin.conf and will suggest some good starting values.

For Windows users, access the RIchCoin.conf file in ```c:\Users\XXX\AppData\Roaming\RIchCoin``` you created earlier.

For Linux users, type:

```nano ~/.RIchCoin/RIchCoin.conf```

For Mac users, type:

```nano ~/Library/Application\ Support\RIchCoin\RIchCoin.conf```

Paste the output from the `RIchCoind` command into the Richcoin.conf like this: (It is recommended to change the password to something unique.)

    rpcuser=RIchcoinRPCusername
    rpcpassword=RIchCoinDevPassword

Add `daemon=1`.

**Optional**: Add `rpcport=20203`, `port=2342`, or `algo=groestl` to the configuration file.

Your config may look something like this:

    rpcuser=Richcoinrpcusername
    rpcpassword=85CpSuCNvDcYsdQU8w621mkQqJAimSQwCSJL5dPT9wQX
    rpcport=20203
    port=2342
    daemon=1
    algo=groestl

Mac and Linux: Exit the Richcoin.conf by pressing `ctrl + x` on your keyboard then pressing `y` and hitting enter. This should have created a Richcoin.conf file with what you just added.

Type: 
```RIchCoind.exe``` (for Windows users)
```./RIchCoind``` (for Mac and Linux users)
and your RIchCoin daemon should start! Huzzah!


To check the status of how much of the blockchain has been downloaded (aka synced) type:
`RIchCoind.exe getinfo` (for Windows users)
`./RIchCoind getinfo` (for Mac and Linux users)

"Solo-mining" from your wallet?
----------
The wallet has a built-in CPU miner. You need to specify the algorithm in your .conf file and set the "gen" flag. For instance, in the configuration specify ```gen=1```.


Using different mining algorithms
----------
To use a specific mining algorithm use the `algo` switch in your configuration file (.conf file) or from the command line (like this `--algo=x17`) Here are the possible values:

    algo=x17
    algo=scrypt
    algo=groestl
    algo=lyra
    algo=blake


Thanks
----
Special Thanks to:
* RIchCoin Devs
* VERGE Devs
* Tor Devs
* i2p Devs
* Bitcoin Devs
* Any others who have helped one way or another


releases.txt
# see http://stackoverflow.com/questions/12343452/how-to-publish-artifacts-in-travis-ci/33109519

To create a release, tag the commit:
   git commit -m 'some commit message'
   git tag -m 'new shiny release' 2.0.0.2 c02aa6e8ae407f3693a2fb26acbd0108d311cd71
   git push
   git push --tags

To delete a tag:
   git tag -d 2.0.0.2

To see tags: ('-n' shows comment, '-l' shows in order)
   git tag -n -l


-- Minimum OS for installation validation
- Mac OSX 10.8 Mountain Lion
- Windows 7
- Ubuntu 14.04
  - For Ubuntu 14.04, you need to run these steps before you can run RIchCoin-qt:
    sudo apt-get install -y libboost-filesystem-dev libboost-program-options-dev libboost-thread-dev libqrencode-dev
    sudo add-apt-repository ppa:bitcoin/bitcoin
    sudo apt-get update
    sudo apt-get install libdb4.8++-dev


go.sh
#!/bin/bash
#// full deployement : run sh go.sh

# generating entropy make it harder to guess the randomness!.
echo "Initializing random number generator..."
random_seed=/var/run/random-seed
# Carry a random seed from start-up to start-up
# Load and then save the whole entropy pool
if [ -f $random_seed ]; then
    sudo cat $random_seed >/dev/urandom
else
    sudo touch $random_seed
fi
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

#Also, add the following lines in an appropriate script which is run during the$

# Carry a random seed from shut-down to start-up
# Save the whole entropy pool
echo "Saving random seed..."
random_seed=/var/run/random-seed
sudo touch $random_seed
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

# Create a swap file

cd ~
if [ -e /swapfile1 ]; then
echo "Swapfile already present"
else
sudo dd if=/dev/zero of=/swapfile1 bs=1024 count=524288
sudo mkswap /swapfile1
sudo chown root:root /swapfile1
sudo chmod 0600 /swapfile1
sudo swapon /swapfile1
fi

# Install dependency

sudo apt-get -y install software-properties-common

sudo add-apt-repository -y ppa:bitcoin/bitcoin

sudo apt-get update

sudo apt-get -y install libcanberra-gtk-module

# Dont need to check if bd is already installed, will override or pass by
#results=$(find /usr/ -name libdb_cxx.so)
#if [ -z $results ]; then
sudo apt-get -y install libdb4.8-dev libdb4.8++-dev
#else
#grep DB_VERSION_STRING $(find /usr/ -name db.h)
#echo "BerkeleyDb will not be installed its already there...."
#fi

sudo apt-get -y install git build-essential libtool autotools-dev autoconf automake pkg-config libssl-dev libevent-dev bsdmainutils git libprotobuf-dev protobuf-compiler libqrencode-dev

sudo apt-get -y install libqt5gui5 libqt5core5a libqt5webkit5-dev libqt5dbus5 qttools5-dev qttools5-dev-tools

sudo apt-get -y install libminiupnpc-dev

# Keep current version of libboost if already present
results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]; then
sudo apt-get -y install libboost-all-dev
else
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`
echo "${red}Libboost will not be installed its already there....${reset}"
grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

sudo apt-get -y install --no-install-recommends gnome-panel

sudo apt-get -y install lynx

sudo apt-get -y install unzip

cd ~

#// Compile Berkeley if 4.8 is not there
if [ -e /usr/lib/libdb_cxx-4.8.so ]
then
echo "BerkeleyDb already present...$(grep --include *.h -r '/usr/' -e 'DB_VERSION_STRING')" 
else
wget http://download.oracle.com/berkeley-db/db-4.8.30.NC.tar.gz 
tar -xzvf db-4.8.30.NC.tar.gz 
rm db-4.8.30.NC.tar.gz
cd db-4.8.30.NC/build_unix 
../dist/configure --enable-cxx 
make 
sudo make install 
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb-4.8.so /usr/lib/libdb-4.8.so
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb_cxx-4.8.so /usr/lib/libdb_cxx-4.8.so
cd ~
sudo rm -Rf db-4.8.30.NC
#sudo ldconfig
fi

#// Check if libboost is present

results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]; then
sudo rm download
     wget https://sourceforge.net/projects/boost/files/boost/1.63.0/boost_1_63_0.zip/download 
     unzip -o download
     cd boost_1_63_0
	sh bootstrap.sh
	sudo ./b2 install
	cd ~
	sudo rm download 
	sudo rm -Rf boost_1_63_0
	#sudo ln -s $(dirname "$(find /usr/ -name libboost_chrono.so)")/lib*.so /usr/lib
	sudo ldconfig
        #sudo rm /usr/lib/libboost_chrono.so
else
     echo "Libboost found..." 
     grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

#// Clone files from repo, Permissions and make

git clone https://github.com/RIchCoin/shield
cd RIchCoin
sudo sh autogen.sh
chmod 777 ~/RIchCoin/share/genbuild.sh
chmod 777 ~/RIchCoin/src/leveldb/build_detect_platform



grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"

sudo rm wrd01.txt
sudo rm wrd00.txt
sudo rm words
find /usr/ -name libboost_chrono.so > words
split -dl 1 --additional-suffix=.txt words wrd



if [ -e wrd01.txt ]; then
echo 0. $(cat wrd00.txt)
echo 1. $(cat wrd01.txt)
echo 2. $(cat wrd02.txt)
echo 3. $(cat wrd03.txt)
echo -n "Choose libboost library to use(0-3)?"
read answer
else
echo "There is only 1 libboost library present. We choose for you 0"
answer=0
fi

echo "You have choosen $answer"

if [ -d /usr/local/BerkeleyDB.4.8/include ]
then
./configure CPPFLAGS="-I/usr/local/BerkeleyDB.4.8/include -O2" LDFLAGS="-L/usr/local/BerkeleyDB.4.8/lib" --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)")
echo "Using Berkeley Generic..."
else
./configure --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)")
echo "Using default system Berkeley..."
fi

make -j$(nproc)

if [ -e ~/RIchCoin/src/qt/RIchCoin-qt ]; then
#sudo apt-get -y install pulseaudio
#sudo apt-get -y install portaudio19-dev
# synthetic voice 
#cd ~
#wget https://sourceforge.net/projects/espeak/files/espeak/espeak-1.48/espeak-1.48.04-source.zip/download
#unzip -o download
#cd espeak-1.48.04-source/src
#cp portaudio19.h portaudio.h
#make
#cd ~
for i in 528 1000 1600 2000 3000
do
pactl load-module module-sine frequency=$i > /dev/null
sleep 0.1
done
sleep 1
pactl unload-module module-sine
pactl upload-sample complet test2
openssl rand -hex 4096 | padsp tee /dev/audio > /dev/null
sleep 1.3
pactl play-sample test2
#espeak mission,complete
sudo strip ~/RIchCoin/src/RIchCoind
sudo strip ~/RIchCoin/src/qt/RIchCoin-qt
sudo make install
else
echo "Compile fail not RIchCoin-qt present"
fi

cd ~

#// Create the config file with random user and password

mkdir ~/.RIchCoin
cp ~/.RIchCoin/RIchCoin.conf ~/.RIchCoin/RIchCoin.bak
echo "rpcuser="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26 ; echo '') '\n'"rpcpassword="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26 ; echo '') '\n'"rpcport=20203" '\n'"port=21203" '\n'"daemon=1" '\n'"listen=1" '\n'"server=1" '\n'"addnode=103.18.40.125:21203" '\n'"addnode=104.131.144.82:21203" '\n'"addnode=138.197.68.130:21203" '\n'"addnode=144.76.167.66:21203" '\n'"addnode=152.186.36.86:21203" '\n'"addnode=159.203.121.202:21203" '\n'"addnode=172.104.157.38:21203" '\n'"addnode=192.99.7.127:21203" '\n'"addnode=219.89.84.46:21203" '\n'"addnode=45.32.129.168:21203" '\n'"addnode=45.55.59.206:21203" '\n'"addnode=46.4.64.68:21203" '\n'"addnode=51.15.46.1:21203" '\n'"addnode=52.9.109.214:21203" '\n'"addnode=66.55.64.183:21203" '\n'"addnode=67.167.207.164:21203" '\n'"addnode=78.46.190.152:21203"> ~/.RIchCoin/RIchCoin.conf

#// Extract http link, download blockchain and install it.

echo -n "Do you wish to download the complete RIchCoin Blockchain (y/n)?"
read answer
if echo "$answer" | grep -iq "^y" ;then
    sudo rm RIchCoin-Blockchain*.zip
    until [ -e RIchCoin*.zip ]
    do
    sleep 1
    echo "wget" $(lynx --dump --listonly https://RIchCoin.de | grep -o "https:*.*zip") > link.sh
    sleep 1
    sh link.sh
    done
    
    #checksum
    sudo rm blockchain
    wget https://www.RIchCoin.github.io/checksums/blockchain
    md5sum RIchCoin-Blockchain*.zip > md5
    checksum="$(grep $(cat md5) blockchain)"
    if [ -z "$checksum" ];
    then
    echo "Warning: MD5 is not matching"
    else
    echo "MD5 is matching...Success"
    fi
    
    unzip -o RIchCoin-Blockchain*.zip -d ~/.RIchCoin
    sudo rm RIchCoin-Blockchain*.zip
else
 echo "Blockchain will not be installed sync may be long"   
fi

# Create Icon on Desktop and in menu

sudo cp ~/RIchCoin/src/qt/res/icons/RIchCoin.png /usr/share/icons/
echo '#!/usr/bin/env xdg-open''\n'"[Desktop Entry]"'\n'"Version=1.0"'\n'"Type=Application"'\n'"Terminal=false"'\n'"Icon[en]=/usr/share/icons/RIchCoin.png"'\n'"Name[en]=RIchCoin"'\n'"Exec=RIchCoin-qt"'\n'"Name=RIchCoin"'\n'"Icon=/usr/share/icons/RIchCoin.png"'\n'"Categories=Network;Internet;" > ~/Desktop/RIchCoin.desktop
sudo chmod +x ~/Desktop/RIchCoin.desktop
sudo cp ~/Desktop/RIchCoin.desktop /usr/share/applications/RIchCoin.desktop
sudo chmod +x /usr/share/applications/RIchCoin.desktop

# Erase all RIchCoin compilation directory , cleaning

cd ~
#sudo rm -Rf ~/RIchCoin

#// Start RIchCoin

RIchCoin-qt
cp ~/.RIchCoin/wallet.dat ~/RIchCoinwallet.bak
